src/disc_bsd.c

```diff
@@ -23,29 +23,37 @@
 
 --------------------------------------------------------------------------- */
 
+#include <assert.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
-#include <sys/types.h>
 #include <sys/cdio.h>
 #include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#if !defined(__FreeBSD__)
+#include <util.h>
+#endif
 
 #include "discid/discid_private.h"
 #include "unix.h"
 
-int mb_disc_unix_read_toc_header(int fd, mb_disc_toc *toc) {
-	struct ioc_toc_header th;
-
-	int ret = ioctl(fd, CDIOREADTOCHEADER, &th);
+#define MAX_DEV_LEN 15
 
-	if ( ret < 0 )
-		return 0; /* error */
-
-	toc->first_track_num = th.starting_track;
-	toc->last_track_num = th.ending_track;
-
-	return 1;
+static int get_device(int n, char* device_name, size_t device_name_length) {
+#if !defined(__FreeBSD__) /* /dev/rcdNX, where X is the letter for the raw partition */
+	snprintf(device_name, device_name_length, "/dev/rcd%d%c", n - 1, 'a' + getrawpartition());
+#else /* on FreeBSD it's just /dev/cdN */
+	snprintf(device_name, device_name_length, "/dev/cd%d", n - 1);
+#endif
+	return mb_disc_unix_exists(device_name);
 }
 
+/* not actually used for *BSD, but needed by unix.c */
```

Comment:
OK, I'll do that then. Would the looping version then be duplicated in each disc_xxx.c? It'll be small enough for that not to matter too much, I suppose.

I suppose the ideal would be a struct with function pointers for read_toc, read_toc_entries, read_toc_entry, etc) - that would allow overriding one for a given platform without worrying about linking duplicate names.
For example, the Linux TOC reading could be changed to use SG_IO to read the entire TOC (header + entries) in one go, but that would currently be tricky because disc_linux can't override mb_disc_unix_read_toc).